### PR TITLE
[BENCHMARKS 1/x] Add base framework and example

### PR DESCRIPTION
**Pull request by sbuckfelder**
Signed-off-by: Scott Buckfelder <buckscot@amazon.com>

This is the first of a series of PRs that builds out performance benchmark testing for soci-snapshotter.

*Description of changes:*
This PR does 5 main things:
- it adds a main file with an example benchmark test for benchmark that is used to compile a benchmark binary.  These examples will be removed in future PRs, but do provide an example of functionality like passing in arguments (imagine different test images)
- adds a framework package (later envisioned to live in containerd upstream project)
- adds a Makefile target to build and run the benchmark binary
- updates `go.mod` to include a statistics package and the framework package
- updates `.gitignore` to exclude the benchmarking output

The way this is envisioned to be used is that the `framework` package will include anything for the benchmark framework or containerd related.  All of the soci specific logic will live in the top level benchmark directory / package.  The main concepts here are the **BenchmarkTestDriver**, this is a struct that houses the function you are benchmarking and stores all of the results for that function, and the **BenchmarkFramework**, this is the struct that has the list of  BenchmarkTestDrivers and runs them.  The benchmarks are run using the golang testing package, but the key difference is that the tests are run individually.  That allows us to run custom statistics on them (the default golang benchmarking statistics only includes averages).  Finally, everything is written out to a json file.

*Testing performed:*
`make && make test && make check`
`make benchmarks` results -->
```
{
 "commit": "474b26470659476af010cd515b7088d76c930e17",
 "benchmarkTests": [
  {
   "testName": "TwoSecondSleep",
   "numberOfTests": 10,
   "testsRun": 0,
   "stdDev": 0.0005865096459530768,
   "mean": 2.0006229670000004,
   "min": 2.000099263,
   "pct25": 2.0000997145,
   "pct50": 2.00015568,
   "pct75": 2.001075953,
   "pct90": 2.001147467,
   "max": 2.001834204
  },
  {
   "testName": "TwoSecondRandSleep",
   "numberOfTests": 10,
   "testsRun": 0,
   "stdDev": 0.5001335439728879,
   "mean": 0.5001444039,
   "min": 0.000010809,
   "pct25": 0.000010888,
   "pct50": 0.000011036,
   "pct75": 1.0000986754999999,
   "pct90": 1.000149617,
   "max": 1.000995927
  },
  {
   "testName": "ThreeSecondSleep",
   "numberOfTests": 10,
   "testsRun": 0,
   "stdDev": 0.001095211523873395,
   "mean": 3.0008177132,
   "min": 3.000097101,
   "pct25": 3.0001005155,
   "pct50": 3.000116305,
   "pct75": 3.0011728705,
   "pct90": 3.002226273,
   "max": 3.002962556
  }
 ]
}
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 /out
 *.db
 .vscode/
+/benchmark/bin
```
Comment (sbuckfelder) [inline .gitignore:4]: location of the benchmark binary

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,5 @@
...
+/benchmark/output.json
```
Comment (sbuckfelder) [inline .gitignore:5]: contains the benchmarking results

File: Makefile [hunk 3]
```diff
@@ -126,3 +127,7 @@ integration:
 	@echo "$@"
 	@echo "SOCI_SNAPSHOTTER_PROJECT_ROOT=$(SOCI_SNAPSHOTTER_PROJECT_ROOT)"
 	@GO111MODULE=$(GO111MODULE_VALUE) SOCI_SNAPSHOTTER_PROJECT_ROOT=$(SOCI_SNAPSHOTTER_PROJECT_ROOT) ENABLE_INTEGRATION_TEST=true go test $(GO_TEST_FLAGS) -v -timeout=0 ./integration
+
+benchmarks:
```
Comment (rdpsin) [inline Makefile:131]: Should add this target to .PHONY.

Comment (wmesard) [reply Makefile:131]: https://www.gnu.org/software/make/manual/make.html#Phony-Targets

File: benchmark/framework/framework.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	Pct75         float64          `json:"pct75"`
+	Pct90         float64          `json:"pct90"`
+	Max           float64          `json:"max"`
+}
+
+func (frame *BenchmarkFramework) Run() {
+	testing.Init()
+	flag.Set("test.benchtime", "1x")
+	flag.Parse()
+	for i := 0; i < len(frame.Drivers); i++ {
```
Comment (vkuzniet) [inline benchmark/framework/framework.go:55]: Maybe use `range` instead?

File: benchmark/framework/framework.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+		for j := 0; j < testDriver.NumberOfTests; j++ {
+			res := testing.Benchmark(testDriver.TestFunction)
+			testDriver.TestTimes = append(testDriver.TestTimes, res.T.Seconds())
+		}
+		testDriver.calculateStats()
+	}
+
+	json, err := json.MarshalIndent(frame, "", " ")
+	if err != nil {
+		fmt.Printf("JSON Marshalling Error: %v\n", err)
```
Comment (vkuzniet) [inline benchmark/framework/framework.go:66]: Do we want to return the `error` instead?

File: benchmark/framework/framework.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	err = os.WriteFile(frame.OutputFile, json, 0644)
+	if err != nil {
+		fmt.Printf("WriteFile Error: %v\n", err)
+	}
+}
+
+func (driver *BenchmarkTestDriver) calculateStats() {
```
Comment (vkuzniet) [inline benchmark/framework/framework.go:74]: For `calculateStats`, is there any value in incomplete stats? Maybe print something out and pass error through?

**Pull request closed without merging**